cpp/src/reductions/simple.cuh

```diff
@@ -50,33 +50,46 @@ namespace detail {
  * @tparam Op           the operator of cudf::reduction::op::
 
  * @param col Input column of data to reduce
+ * @param init Optional initial value of the reduction.
  * @param stream Used for device memory operations and kernel launches.
  * @param mr Device memory resource used to allocate the returned scalar's device memory
  * @return Output scalar in device memory
  */
 template <typename ElementType, typename ResultType, typename Op>
 std::unique_ptr<scalar> simple_reduction(column_view const& col,
+                                         std::optional<const scalar*> init,
                                          rmm::cuda_stream_view stream,
                                          rmm::mr::device_memory_resource* mr)
 {
   // reduction by iterator
   auto dcol      = cudf::column_device_view::create(col, stream);
   auto simple_op = Op{};
 
+  // Cast initial value
+  std::optional<ResultType> initial_value;
+  if (init.has_value() && init.value()->is_valid()) {
+    using ScalarType = cudf::scalar_type_t<ElementType>;
+    auto input_value = static_cast<const ScalarType*>(init.value())->value(stream);
+    initial_value    = static_cast<ResultType>(input_value);
+  } else {
+    initial_value = std::nullopt;
+  }
```

Comment:
Use an immediately-invoked function expression (IIFE) to avoid the "initialize, then modify" pattern. I think this is what you need:
```suggestion
  // Cast initial value
  std::optional<ResultType> const initial_value = [&] {
    if (init.has_value() && init.value()->is_valid()) {
      using ScalarType = cudf::scalar_type_t<ElementType>;
      auto input_value = static_cast<const ScalarType*>(init.value())->value(stream);
      return static_cast<ResultType>(input_value);
    }
    return std::nullopt;
  }();
```

Reply:
I could see this being written as a detail function `get_initial_value(init, stream)` because of how many times this logic is repeated later in the file.